File: me_daily/2025/10/1031/cf pinely round5 div12/c/cf pinely div12 c.py
from collections import defaultdict
from typing import List
import sys
from collections import Counter
input = lambda: sys.stdin.readline().strip()
II = lambda: int(input())
LII = lambda: list(map(int, input().split()))
SI = lambda: input()
LSI = lambda: input().split()

t = II()

def solve_simple():
    n, X = map(int, input().split())
    prices = list(map(int, input().split()))
    
    total = sum(prices)
    levels = total // X  # 总共能达到的等级数
    
    # 按价格降序排列
    prices.sort(reverse=True)
    
    # 前levels个最大的价格就是奖励积分
    bonus_points = sum(prices[:levels])
    
    # 构造一个有效的购买顺序
    # 策略：小的物品先买，大的物品在关键时刻买
    small_items = prices[levels:]
    large_items = prices[:levels]
    
    result = []
    current_sum = 0
    large_idx = 0
    
    for level in range(1, levels + 1):
        target = level * X
        
        # 添加小物品直到接近目标
        while current_sum < target - large_items[large_idx] and small_items:
            item = small_items.pop()
            result.append(item)
            current_sum += item
        
        # 添加大物品触发等级增加
        result.append(large_items[large_idx])
        current_sum += large_items[large_idx]
        large_idx += 1
    
    # 添加剩余物品
    result.extend(small_items)
    
    return bonus_points, result

for _ in range(t):
    p, r = solve_simple()
    print(p)
    print(*r)

File: me_daily/2025/10/1031/cf pinely round5 div12/c/cf pinely div12 c.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

var reader = bufio.NewReader(os.Stdin)
var writer = bufio.NewWriter(os.Stdout)

func readLine() string {
	s, _ := reader.ReadString('\n')
	return strings.TrimSpace(s)
}

func readInts() []int {
	fields := strings.Fields(readLine())
	res := make([]int, len(fields))
	for i, f := range fields {
		res[i], _ = strconv.Atoi(f)
	}
	return res
}

func sum[E ~int | ~float64](s []E) E {
	var total E
	for _, v := range s {
		total += v
	}
	return total
}

func main() {
	defer writer.Flush() //main() 结束时自动调用
	t, _ := strconv.Atoi(readLine())

	for ; t > 0; t-- {
		header := readInts()
		n, x := header[0], header[1]
		prices := readInts()
		sumP := sum(prices)
		levels := sumP / x // 防溢，用int64算levels
		sort.Slice(prices, func(i, j int) bool {
			return prices[i] > prices[j]
		})
		var curS int = 0
		smallItems := make([]int, len(prices[levels:]))
		copy(smallItems, prices[levels:])
		largeItems := make([]int, levels)
		copy(largeItems, prices[:levels])
		var res []int
		largeIdx := 0
		bonusPoints := int(sum(prices[:levels]))
		for level := 1; level <= levels; level++ {
			target := level * x
			for curS < target-largeItems[largeIdx] && len(smallItems) > 0 {
				item := smallItems[len(smallItems)-1]
				smallItems = smallItems[:len(smallItems)-1]
				res = append(res, item)
				curS += item
			}
			res = append(res, largeItems[largeIdx])
			curS += largeItems[largeIdx]
			largeIdx++
		}
		res = append(res, smallItems...)
		fmt.Fprintln(writer, bonusPoints)

		// 输出res，空格分隔
		if len(res) > 0 {
			out := make([]string, len(res))
			for i, v := range res {
				out[i] = strconv.Itoa(v)
			}
			fmt.Fprintln(writer, strings.Join(out, " "))
		} else {
			fmt.Fprintln(writer)
		}
	}
}
